### PR TITLE
[WIP] Optimistic UI

### LINKED ISSUE #287: Optimistic updates
Client-side optimistic updates should work pretty seamlessly if there's a way to tell the mutation "here's the data to use until the response comes back".  The trick will be figuring out how to get that into the current syntax and deciding how it gets undone in the event of a failure, since a failed server call probably can't be expected to reliably undo the optimistic update by sending down the opposite change (seeing how it failed, after all).

### PR DESCRIPTION
**Pull request by davidwoody**
TODO:

- [x] Use `merge` instead of `assign` to combine optimistic data
- [x] Optimistic reducer should use the exact same logic for `data` as a real mutation result
- [x] Update CHANGELOG.md with your change
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Make sure all of the significant new logic is covered by tests

closes #287

### CODE REVIEW

**Comment by abhiaiyer91** [comment]
How does Optimistic Data get reset or trashed? I might have missed that in the PR?

**Comment by stubailo** [comment]
We should also make sure that the optimistic result is applied using the merged real and optimistic data, so that you can apply multiple optimistic updates on top of each other.

**Comment by abhiaiyer91** [comment]
Bump on my question @davidwoody

**Comment by abhiaiyer91** [comment]
What is the lifecycle of optimistic state?

**Comment by davidwoody** [comment]
Does the logic in the optimistic reducer make sense to you? Basically on init it simulates the result action and on the last result action it throws away the optimistic cache. It waits until the last outstanding result action to avoid the flickering issue I was explaining in Slack the other day.

**Comment by davidwoody** [comment]
I wasn't sure how best to "throw away" the one optimistic result after the server result became available if there was still another optimistic result applied to the cache and waiting for the server responses. So I just wait to throw away the optimistic results until all outstanding sever mutations resolve.

**Comment by stubailo** [comment]
> So I just wait to throw away the optimistic results until all outstanding sever mutations resolve.

This seems like a good approach for now.

**Comment by stubailo** [comment]
Hey, heads up @davidwoody I rebased on master and force pushed

**Comment by davidwoody** [comment]
@stubailo Got it. Also, it looks like the travis builds are failing because the gzipped size of index.min.js exceeds 34kb... I'm guessing we can just increase that from 34kb to 36kb?

https://travis-ci.org/apollostack/apollo-client/jobs/141736810#L1343

**Comment by Slava** [comment]
Plan for new tests for this functionality:

- [ ] reuse the mutations tests structure for different types, that the result is optimistically reflected synchronously:
    - [ ] correctly integrates a basic object at the end
    - [ ] deletes object from array and store
    - [ ] runs the custom reducer
    - [ ] deletes an object from array but not store
- [ ] test error handling
   - [ ] one optimistic update in a stack
   - [ ] multiple in a stack, one fails (in the beginning or middle)
- [ ] two mutations running concurrently and returning in some order. every intermediate state is validated

**Comment by davidwoody** [comment]
@Slava I believe the `data` reducer needs the `mutations` state created during the `APOLLO_MUTATION_INIT` action here: https://github.com/apollostack/apollo-client/pull/336/files#diff-906bf6551e27e8a53f442ca0d5d7d359R36

**Comment by Slava** [comment]
Added some tests that demonstrate the problem of not passing the previous data state (with optimistic updates so far).

**Comment by Slava** [comment]
Completed the test suite planned by me and @stubailo awhile ago.

File: CHANGELOG.md [hunk 1]
```diff
@@ -44,6 +44,8 @@ Expect active development and potentially significant breaking changes in the `0
 - Deprecate `apollo-client/gql` for `graphql-tag` and show a meaningful warning when importing
   `apollo-client/gql`
 
+- Allow `client.mutate` to accept an `optimisticResponse` argument to update the cache immediately, then after the server responds replace the `optimisticResponse` with the real response. [Issue #287](https://github.com/apollostack/apollo-client/issues/287) [PR #336](https://github.com/apollostack/apollo-client/pull/336)
```
Comment (stubailo) [inline CHANGELOG.md]: This is not in the right location, should be moved to the top.

File: src/optimistic-data/store.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    const optimisticState = {
+      data: patch,
+      mutationId: action.mutationId,
+    };
+
+    const newState = [...previousState, optimisticState];
+
+    return newState;
+  } else if ((isMutationErrorAction(action) || isMutationResultAction(action))
+               && action.optimisticResponse) {
```
Comment (stubailo) [inline src/optimistic-data/store.ts]: Seems odd that the result/error actions need to include an `optimisticResponse` field. Perhaps we can look into the `previousState` to see if the mutation id matches?

**Comment by stubailo** [comment]
Can we add a test that makes sure that the optimistic patch is small, and doesn't include objects that weren't changed? That seems like one thing that would be easy to break in future changes.

File: src/optimistic-data/store.ts (hunk 2 above)
Comment (Slava) [inline src/optimistic-data/store.ts]: Good point, will do.

**Comment by Slava** [comment]
Yay we merged! Thanks @davidwoody for the good work on this feature.

**Comment by davidwoody** [comment]
Woo hoo! Thanks @Slava for the great suggestions and work on all the tests!

**Comment by stubailo** [comment]
Great work everyone!

**Comment by twelvearrays** [comment]
This is very exciting! Are there any examples available of using optimisticResponse with a client mutation?

**Comment by Slava** [comment]
@twelvearrays we are still in the process of designing a good API and ironing out the details of what is required to be passed and what is not. Once that's done better docs should come along!

**Comment by twelvearrays** [comment]
@Slava Thanks! Looking forward to it. Keep up the great work.

**Comment by helfer** [comment]
1. The optimistic response should basically be what you expect the server to return if everything works out as expected. That also means you have to predict what fragments will apply. It could be that there's a bug in the logic that checks that the shape of the optimisticResponse fits the query, but then that would probably throw an error on the actual response as well.

If I understand correctly, you're saying that using an optimisticResponse that matches the Error fragment doesn't work? If so, that might be a bug.

2. optimistic updates should affect all queries, and their subscriptions should trigger if the optimistic update affected their data. If that's not the case here, it could be a bug, or it could be that it's not actually mapped to the same place in the store. It would be great to have a minimal reproduction or a failing test case to figure out what exactly is happening.

**Comment by leebenson** [comment]
Thanks @helfer. I'll try to reproduce with a minimal test, and post the results here.

**Pull request merged**